hal/inc/watchdog_hal.h

```diff
@@ -1,41 +1,77 @@
-/**
- ******************************************************************************
- * @file    watchdog_hal.h
- * @author  Matthew McGowan
- * @version V1.0.0
- * @date    25-Sept-2014
- * @brief
- ******************************************************************************
-  Copyright (c) 2013-2015 Particle Industries, Inc.  All rights reserved.
-
-  This library is free software; you can redistribute it and/or
-  modify it under the terms of the GNU Lesser General Public
-  License as published by the Free Software Foundation, either
-  version 3 of the License, or (at your option) any later version.
-
-  This library is distributed in the hope that it will be useful,
-  but WITHOUT ANY WARRANTY; without even the implied warranty of
-  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
-  Lesser General Public License for more details.
-
-  You should have received a copy of the GNU Lesser General Public
-  License along with this library; if not, see <http://www.gnu.org/licenses/>.
- ******************************************************************************
+/*
+ * Copyright (c) 2022 Particle Industries, Inc.  All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation, either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, see <http://www.gnu.org/licenses/>.
  */
 
 #ifndef WATCHDOG_HAL_H
 #define	WATCHDOG_HAL_H
 
 #include <stdint.h>
 #include <stdbool.h>
+#include "hal_platform.h"
 
 #ifdef	__cplusplus
 extern "C" {
 #endif
 
-void HAL_Watchdog_Init(uint32_t timeout_tick_ms);
-bool HAL_watchdog_reset_flagged();
-void HAL_Notify_WDT();
+#if HAL_PLATFORM_HW_WATCHDOG
+
+#define HAL_WATCHDOG_VERSION    1
+
+typedef void (*hal_watchdog_on_expired_callback_t)(void);
+
+typedef enum hal_watchdog_capability_t {
+    WATCHDOG_CAPS_NONE              = 0x00,
+    WATCHDOG_CAPS_RESET             = 0x01,
+    WATCHDOG_CAPS_INT               = 0x02,
+    WATCHDOG_CAPS_RECONFIGURABLE    = 0x04,
+    WATCHDOG_CAPS_STOPPABLE         = 0x08,
+    WATCHDOG_CAPS_ALL               = 0xFFFFFFFF
+} hal_watchdog_capability_t;
+
+typedef struct hal_watchdog_config_t {
```

Comment:
What is the relationship between the watchdog timeout and the callback?  Is there an option to hard reset the system or get a callback?

Reply:
I see the option in the Wiring layer but nothing to point back here https://github.com/particle-iot/device-os/pull/2595/files#diff-57620cd25673be85a34ad562b345fa78794b392eceeb4f3c9d2395d2d5900e4bR122

Reply:
Some watchddogs support interrupt before reset, so when it timeout, callback is available for user to handle something before reset. I would likee to rename `WATCHDOG_CAPS_RESET` to `WATCHDOG_CAPS_SOFT_RESET`, cos normally when watchdog expires it means that some fatal error occurs and a hard reset is required, but maybe for some watchdogs they allow user to do soft-reset in calback and block the hard reset. And a `hard_reset` option can be introduced in the config struct so that when `WATCHDOG_CAPS_SOFT_RESET` is supported, user can decide whether the watchdog need to perform the hard reset. If the soft-reset is not supported, the `hard_reset` will be initialized to `true` forcely.